### ISSUE TITLE
Allow using "require-implements / require-extends" annotations from classes and interfaces scope

### ISSUE THREAD

**Issue by awd-studio** [vimeo/psalm#6043]
These features works great from tratis: https://psalm.dev/r/a9899273ff

At the same time, it'd be cool to have it working in the other members: https://psalm.dev/r/1d300f0c72

**Comment by VincentLanglet** [comment]
You can replace the `require-implements` by 
```
interface C extends A {}
```

And I would say a class should based on an interface, not the opposite.
So having an interface which should "extends" a class does not make sens.

If you want to use method which are in the abstract class B and not in the interface, they should be added to the interface instead. If you want to force a special implementation, that's not the purpose of an interface.

**Comment by awd-studio** [comment]
I'd like to control type intersection. :man_shrugging:

**Comment by VincentLanglet** [comment]
> I'd like to control type intersection. 🤷‍♂️

Do you have a real case usage ? Because I find very strange a non following the php strategy.

**Comment by awd-studio** [comment]
I really have. Not sure I could explain it in a few words, but I can try.. 

So, I have a "projection" interface, that members will implement. 
I have a "projection-record", a thing, that will store "projections". 
I have some kind of "tagger" - a thing, that can add tags to a "projection-record", based on a "projection" it contains (actually on other interfaces of "taggable"). 
I want to leave a little tip, for myself, for each projection-member to implement at least one taggable interface. 

Trying to draw as code:
```php
interface Tags {}

interface Taggable {}

/** 
 * This, how I'd like it wok
 * @psalm-require-implements Taggable 
 */
interface Projection {}

final class ProjectionRecord {
    public Projection $projection;
    public Tags $tags;
}

interface Tagger {
    public function tag(Taggable $taggable): Tags;
}

final class ProjectionRecordFactory {
    private Tagger $tagger;

    public function build(Projection $projection): ProjectionRecord {
        // Now, I check if a projection implements Taggable here
        $tags = $this->tagger->tag($projection);

        return new ProjectionRecord($projection, $tags);
    }
}
```

Hope, it enough clear.

**Comment by VincentLanglet** [comment]
Wouldn't changing
```
/** 
 * This, how I'd like it wok
 * @psalm-require-implements Taggable 
 */
interface Projection {}
```
to
```
interface Projection extends Taggable {}
```
solving exactly the use case ?

If you expect a projection to implements Taggable, the interface should extends Taggable IMHO.

**Comment by awd-studio** [comment]
Not really. It doesn't have to implement Taggable, from a perspective of the Projection. Taggable is in different context. It's rather an applicational thing - in the app they should work together. `@psalm-require-implements Taggable` needs as a meta-guaranty for a final app, not for a context.

**Comment by weirdan** [comment]
There's a case for built-in interfaces. Anything implementing `Throwable` has to extend `Exception` or `Error`, anything implementing `DateTimeInterface` has to extend from `DateTime` or `DateTimeImmutable`, and I think same goes for `Iterator`.

For code supporting PHP5 (or just inherited with no modifications) exception case is also useful, as there was no `Throwable` to extend your interfaces from.
